File: ruby/play.rb
require_relative 'world'

class Play
  WORLD_WIDTH  = 150
  WORLD_HEIGHT = 40
  CLEAR_SCREEN = "\x1b[?2026h\x1b[H\x1b[2J"
  SHOW_SCREEN  = "\x1b[?2026l"

  def self.run
    world = World.new(
      width: WORLD_WIDTH,
      height: WORLD_HEIGHT,
    )

    minimal = ENV['MINIMAL'] == '1'

    unless minimal
      print world.render, "\n"
    end

    total_tick = 0
    lowest_tick = Float::INFINITY
    total_render = 0
    lowest_render = Float::INFINITY

    loop do
      tick_start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
      world.dotick
      tick_finish = Process.clock_gettime(Process::CLOCK_MONOTONIC)
      tick_time = (tick_finish - tick_start)
      total_tick += tick_time
      lowest_tick = [lowest_tick, tick_time].min
      avg_tick = (total_tick / world.tick)

      render_start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
      rendered = world.render
      render_finish = Process.clock_gettime(Process::CLOCK_MONOTONIC)
      render_time = (render_finish - render_start)
      total_render += render_time
      lowest_render = [lowest_render, render_time].min
      avg_render = (total_render / world.tick)

      unless minimal
        print CLEAR_SCREEN
      end

      puts format(
        "#%d - World Tick (L: %.3f; A: %.3f) - Rendering (L: %.3f; A: %.3f)",
        world.tick,
        _f(lowest_tick),
        _f(avg_tick),
        _f(lowest_render),
        _f(avg_render)
      )

      unless minimal
        puts rendered + SHOW_SCREEN
      end
    end
  end

  def self._f(value)
    # seconds -> milliseconds
    value * 1_000
  end
  private_class_method :_f
end

Play.run

File: typescript/play.ts
import { World } from './world.js'

// @ts-expect-error Needed for Boa runtime
globalThis.performance ??= {
  now: () => Date.now(),
}

class Play {
  static readonly #WORLD_WIDTH = 150
  static readonly #WORLD_HEIGHT = 40
  static readonly #CLEAR_SCREEN = "\x1b[?2026h\x1b[H\x1b[2J"
  static readonly #SHOW_SCREEN = "\x1b[?2026l"

  public static run() {
    const world = new World(
      Play.#WORLD_WIDTH,
      Play.#WORLD_HEIGHT,
    )

    const minimal =
      typeof Deno === 'object' ? Deno.env.get('MINIMAL') == '1'
      : typeof process === 'object' ? process.env.MINIMAL == '1'
      : true

    if (!minimal) {
      console.log(world.render())
    }

    let totalTick = 0
    let lowestTick = Infinity
    let totalRender = 0
    let lowestRender = Infinity

    while(true) {
      const tickStart = performance.now()
      world.doTick()
      const tickFinish = performance.now()
      const tickTime = (tickFinish - tickStart)
      totalTick += tickTime
      lowestTick = Math.min(lowestTick, tickTime)
      const avgTick = (totalTick / world.tick)

      const renderStart = performance.now()
      const rendered = world.render()
      const renderFinish = performance.now()
      const renderTime = (renderFinish - renderStart)
      totalRender += renderTime
      lowestRender = Math.min(lowestRender, renderTime)
      const avgRender = (totalRender / world.tick)

      if (!minimal) {
        process.stdout.write(Play.#CLEAR_SCREEN)
      }

      console.log(
        `#${world.tick}` +
        ` - World Tick (L: ${Play.#_f(lowestTick)}; A: ${Play.#_f(avgTick)})` +
        ` - Rendering (L: ${Play.#_f(lowestRender)}; A: ${Play.#_f(avgRender)})`
      )

      if (!minimal) {
        console.log(rendered + Play.#SHOW_SCREEN)
      }
    }
  }

  static #_f(value: number) {
    // milliseconds -> no conversion needed
    return value.toFixed(3)
  }
}

Play.run()
